compiler/luci-interpreter/src/kernels/TransposeConv.cpp

```diff
@@ -0,0 +1,174 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kernels/TransposeConv.h"
+
+#include "kernels/Utils.h"
+
+#include <tensorflow/lite/kernels/internal/reference/reference_ops.h>
+
+#include <stdexcept>
+
+namespace luci_interpreter
+{
+
+namespace kernels
+{
+
+TransposeConv::TransposeConv(const Tensor *outputShape, const Tensor *weights,
+                             const Tensor *inputData, Tensor *output,
+                             const TransposeConvParams &params)
+    : KernelWithParams<TransposeConvParams>(params), _outputShape(outputShape), _weights(weights),
+      _inputData(inputData), _output(output)
+{
+}
+
+void TransposeConv::configure()
+{
+  assert(_outputShape->shape().num_dims() == 1);
+  assert(_inputData->shape().num_dims() == 4);
+  assert(_weights->shape().num_dims() == 4);
+  assert(_inputData->element_type() == DataType::FLOAT32 ||
+         _inputData->element_type() == DataType::U8);
+  assert(_inputData->element_type() == _output->element_type());
+  assert(_inputData->shape().dim(3) == _weights->shape().dim(3));
+  Shape im2col_shape(2);
+  im2col_shape.dim(0) = _inputData->shape().dim(1) * _inputData->shape().dim(2);
+  im2col_shape.dim(1) =
+      _weights->shape().dim(0) * _weights->shape().dim(1) * _weights->shape().dim(2);
+  _im2col =
+      std::make_unique<Tensor>(_inputData->element_type(), im2col_shape, AffineQuantization{}, "");
+  if (_inputData->element_type() == DataType::U8)
+  {
+    _scratch_tensor =
+        std::make_unique<Tensor>(DataType::S32, _output->shape(), AffineQuantization{}, "");
+    double real_multiplier = 0.0;
+    const double input_product_scale = _inputData->scale() * _weights->scale();
+    assert(input_product_scale >= 0);
+    real_multiplier = input_product_scale / _output->scale();
+    int exponent;
+    quantizeMultiplier(real_multiplier, &_output_multiplier, &exponent);
+    _output_shift = -exponent;
+    _output_activation_min = std::numeric_limits<uint8_t>::min();
+    _output_activation_max = std::numeric_limits<uint8_t>::max();
+  }
+
+  int dims = _outputShape->shape().dim(0);
+  Shape output_shape(dims);
+  if (_inputData->element_type() == DataType::FLOAT32)
+  {
+    const float *shape_data = getTensorData<float>(_outputShape);
+    for (int i = 0; i < dims; i++)
+      output_shape.dim(i) = shape_data[i];
+  }
+  else
+  {
+    const uint8_t *shape_data = getTensorData<uint8_t>(_outputShape);
```

Comment:
This looks strange. The type of shape depends on the type of input?

Reply:
Oh, Sorry! I will Fix it!